File: Python/pythagorean.py
# A simple code to practice Control Structures solving the missing side
# of a right triangle using the Pythagorean theorem

from math import sqrt


def main():
    leg_one = set_side("first leg")
    leg_two = set_side("second leg", leg_one)
    hypotenuse = set_side("hypotenuse", leg_one, leg_two)
    missing_side(leg_one, leg_two, hypotenuse)
    continue_program()


def set_side(side, leg_one=None, leg_two=None):
    while True:
        if leg_one == "x" or leg_two == "x":
            side_value = input("Enter {}: ".format(side))
        else:
            side_value = input("Enter {} (if it is the missing side enter x): "
                               .format(side))
        if side_value != "x":
            try:
                return float(side_value)
            except ValueError:
                print("Please, enter a proper value")
        elif side_value == "x" and leg_one != "x" and leg_two != "x":
            return side_value
        else:
            print("Please, enter a proper value")


def missing_side(leg_one, leg_two, hypotenuse):
    _missing_side = pythagorean(leg_one, leg_two, hypotenuse)
    if leg_one == "x":
        print("Leg one is: {0:.3f}".format(_missing_side))
    elif leg_two == "x":
        print("Leg two is: {0:.3f}".format(_missing_side))
    else:
        print("Hypotenuse is: {0:.3f}".format(_missing_side))


def pythagorean(leg_one, leg_two, hypotenuse):
    if leg_one == "x":
        _missing_side = sqrt((hypotenuse**2) - (leg_two**2))
        return _missing_side
    elif leg_two == "x":
        _missing_side = sqrt((hypotenuse**2) - (leg_one**2))
        return _missing_side
    _missing_side = sqrt((leg_one**2) + (leg_two**2))
    return _missing_side


def continue_program():
    system_exit = input("Do you want to continue with the program? [Y/n] ")
    system_exit = system_exit.lower()
    if system_exit == "y" or system_exit == "":
        main()
    elif system_exit == "n":
        raise SystemExit
    else:
        print("You entered {}, but it is a wrong command. Good bye!"
              .format(system_exit))
        raise SystemExit


if __name__ == "__main__":
    main()

File: Java/Pythagorean/src/pythagorean/Pythagorean.java
/* A simple code to practice Control Structures solving the missing side
 * of a right triangle using the Pythagorean theorem
 */
package pythagorean;

import java.util.Scanner;

/**
 *
 * @author Javi Palacios <javi@fjp.es>
 */
public class Pythagorean {

    private static boolean continueProgram = true;

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        while (continueProgram) {
            float legOne = setSide(scan);
            float legTwo = setSide(legOne, scan);
            float hypotenuse = setSide(legOne, legTwo, scan);
            missingSide(legOne, legTwo, hypotenuse);
            continueProgram(scan);
        }
    }

    private static float setSide(Scanner scan) {
        return getPossitiveNumberOrZero("Enter leg one ("
                + "if it is the missing side enter 0): ", true, scan);
    }

    private static float setSide(float legOne, Scanner scan) {
        if (legOne == 0) {
            return getPossitiveNumberOrZero("Enter leg two: ", false, scan);
        }
        return getPossitiveNumberOrZero("Enter leg two ("
                + "if it is the missing side enter 0): ", true, scan);
    }

    private static float setSide(float legOne, float legTwo, Scanner scan) {
        if (legOne == 0 || legTwo == 0) {
            return getPossitiveNumberOrZero("Enter hypotenuse: ", false, scan);
        }
        return 0;
    }

    private static float getPossitiveNumberOrZero(String message,
            boolean canBeZero, Scanner scan) {
        float number = -1;
        do {
            try {
                System.out.print(message);
                number = Float.parseFloat(scan.nextLine());
            } catch (NumberFormatException ex) {
                System.out.println("Please, enter only numbers.");
            }
        } while ((canBeZero && number == -1) || (!canBeZero && number < 1));
        return number;
    }

    private static void missingSide(float legOne, float legTwo, float hypotenuse) {
        double missingSide = pythagorean(legOne, legTwo, hypotenuse);
        if (legOne == 0) {
            System.out.printf(String.format("Leg one is: %.3f", missingSide));
        } else if (legTwo == 0) {
            System.out.printf(String.format("Leg two is: %.3f", missingSide));
        } else {
            System.out.printf(String.format("Hypotenuse is: %.3f", missingSide));
        }
    }

    public static double pythagorean(float legOne, float legTwo, float hypotenuse) {
        if (legOne == 0) {
            return Math.sqrt(Math.pow(hypotenuse, 2) - Math.pow(legTwo, 2));
        } else if (legTwo == 0) {
            return Math.sqrt(Math.pow(hypotenuse, 2) - Math.pow(legOne, 2));
        } else {
            return Math.sqrt(Math.pow(legOne, 2) + Math.pow(legTwo, 2));
        }
    }

    private static void continueProgram(Scanner scan) {
        System.out.print("\nDo you want to continue with the program? [Y/n] ");
        String exit = scan.nextLine();
        if ("y".equalsIgnoreCase(exit) || "".equals(exit)) {
            continueProgram = true;
        } else if ("n".equalsIgnoreCase(exit)) {
            continueProgram = false;
        } else {
            System.out.println("You entered a wrong command. Good bye!");
            continueProgram = false;
        }
    }

}
